### ISSUE TITLE
lock file doesn't seem to behave correctly with git-URL-based dependencies

### ISSUE THREAD

**Issue by tom-mazo** [npm/npm#16878]
#### I'm opening this issue because:

  - npm is producing an incorrect install.

#### What's going wrong?

I'm a front-end developer using private Git repos (hosted on BitBucket) and I was blissed to learn that npm v5 is coming with semver versionning for Git URL.
For me this means: I can finally completely replace Bower with npm.

However,  `npm install` now also creates a lock file.

The issue is: this lock file doesn't seem to get updated when I change the version of the private Git repo.

For instance, if I have this in my package.json:

```
{
  "name": "foo",
  .....
  "dependencies" : {
    "jquery": "^3.2.1",
    "private-dep": "git+ssh://git@git/my/private-repo.git#semver:1.3.x"
  }
}
```

This will generate the following lock file:
```
{
........
   "jquery": {
      "version": "3.2.1",
      "resolved": "https://registry.npmjs.org/jquery/-/jquery-3.2.1.tgz",
      "integrity": "sha1-XE2d5lKvbNCncBVKYxu6ErAVx4c="
    },
   "private-dep": {
      "version": "git+ssh://git@git/my/private-
           repo.git#c2877d06a2829d68d23bd567478e8397b8449c2f" 
    },
}
```

It seems to fall back to the old behavior: it points the private repo to a specific commit.

Now, say I update my private repo, like so:
```
{
"dependencies" : {
    "jquery": "^3.2.1",
    "private-dep": "git+ssh://git@git/my/private-repo.git#2.0.0"
  }
}
```

If I run `npm install`, `package-lock.json` will NOT be updated, and the version 2.0.0 of my private dependency will not be installed. Actually, npm will do nothing.
It will only output: `up to date in 3.016s`.

In fact, even if I delete my `node_modules` directory and run again `npm install`, the initial and incorrect version (1.3.x) will still be installed because it has been locked in the lock file, instead of taking the correct version defined in `package.json`.

#### How can the CLI team reproduce the problem?

You can easily reproduce this behavior by running npm install in a project that has Git URLs as dependencies and mess around with the version number.

### supporting information:

 - `npm -v` prints: `5.0.1`
 - `node -v` prints: `v6.10.0`
 - `npm config get registry` prints: `https://registry.npmjs.org/`
 - Windows, OS X/macOS, or Linux?: OS X
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [x] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by zkat** [comment]
This is the "updating package.json doesn't update package-lock" issue.

**Comment by Alexsey** [comment]
for now solution is to delete `package-lock.json` and add to `.npmrc` this line
```
package-lock=false
```
no `package-lock.json` - no problems :)

**Comment by zkat** [comment]
This is fixed as of #17508. It is available in the latest canary, `npmc@5.0.4-canary.13` right now, and will be included in the next npm release next Wednesday. Closing, since the code's already been merged.
